File: frontend/src/types/billing.ts
export interface PlanLimit {
  key: string;
  value: number;
}

export interface BillingPlanStripePrices {
  monthly: boolean;
  yearly: boolean;
}

export interface BillingPlan {
  id: string;
  name: string;
  monthly_price: number;
  yearly_price: number | null;
  features: string[];
  limits: PlanLimit[];
  active: boolean;
  checkout_available?: boolean;
  stripe_prices?: BillingPlanStripePrices;
}

export interface CurrentSubscription {
  plan_id: string;
  plan_name: string;
  status: string;
  billing_cycle: string;
  monthly_price: number;
  yearly_price: number | null;
  auto_renew: boolean;
  cancel_at_period_end: boolean;
  started_at: string | null;
  next_renewal_at: string | null;
  provider: string | null;
  stripe_customer_id: string | null;
}

export interface UpgradePayload {
  plan_id: string;
  billing_cycle: 'monthly' | 'yearly';
}

export interface UpgradeResponse {
  message: string;
  subscription: CurrentSubscription;
}

export interface CheckoutPayload {
  plan_id: string;
  billing_cycle: 'monthly' | 'yearly';
}

export interface CheckoutResponse {
  checkout_url: string;
  session_id: string;
}

export interface PortalResponse {
  portal_url: string;
}

export interface SubscriptionActionResponse {
  message: string;
  subscription: CurrentSubscription;
}

export interface BillingUsageItem {
  metric: string;
  used: number;
  limit: number;
}

export interface BillingUsage {
  items: BillingUsageItem[];
}

export interface BillingHistoryItem {
  id: number;
  event_type: string;
  amount: number;
  currency: string;
  description: string | null;
  stripe_event_id: string | null;
  status: string | null;
  created_at: string;
}

export interface BillingHistory {
  events: BillingHistoryItem[];
}

File: backend/app/schemas/billing.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


class PlanLimit(BaseModel):
    key: str
    value: int


class PlanStripePrices(BaseModel):
    monthly: bool = False
    yearly: bool = False


class PlanRead(BaseModel):
    id: str
    name: str
    monthly_price: float
    yearly_price: float | None = None
    features: list[str]
    limits: list[PlanLimit]
    active: bool = True
    checkout_available: bool = False
    stripe_prices: PlanStripePrices = Field(default_factory=PlanStripePrices)


class CurrentSubscriptionResponse(BaseModel):
    plan_id: str
    plan_name: str
    status: str
    billing_cycle: str
    monthly_price: float
    yearly_price: float | None = None
    auto_renew: bool
    cancel_at_period_end: bool = False
    started_at: datetime | None = None
    next_renewal_at: datetime | None = None
    provider: str | None = None
    stripe_customer_id: str | None = None


class UpgradeRequest(BaseModel):
    """Mantido para compatibilidade interna/administrativa. Fluxo principal usa /billing/checkout."""
    plan_id: str
    billing_cycle: Literal["monthly", "yearly"] = "monthly"


class UpgradeResponse(BaseModel):
    message: str
    subscription: CurrentSubscriptionResponse


# ── Checkout Stripe ─────────────────────────────────────────
class CheckoutRequest(BaseModel):
    plan_id: str
    billing_cycle: Literal["monthly", "yearly"] = "monthly"


class CheckoutResponse(BaseModel):
    checkout_url: str
    session_id: str


# ── Portal Stripe ────────────────────────────────────────────
class PortalResponse(BaseModel):
    portal_url: str


# ── Cancel / Resume ──────────────────────────────────────────
class SubscriptionActionResponse(BaseModel):
    message: str
    subscription: CurrentSubscriptionResponse


# ── Usage ────────────────────────────────────────────────────
class BillingUsageItem(BaseModel):
    metric: str
    used: int
    limit: int


class BillingUsageResponse(BaseModel):
    items: list[BillingUsageItem]


# ── History ──────────────────────────────────────────────────
class BillingHistoryItem(BaseModel):
    id: int
    event_type: str
    amount: float
    currency: str
    description: str | None = None
    stripe_event_id: str | None = None
    status: str | None = None
    created_at: datetime


class BillingHistoryResponse(BaseModel):
    events: list[BillingHistoryItem]
